File: Longest_Word_Dictionary.cpp
// Time Complexity : O(n*l) : TRie construction
// Space Complexity : O(n*l)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
// Your code here along with comments explaining your approach
//1. The trieNode and the insert functions are created to create the trie
//2. Queue is utilized to store the words with each element present from last to first alphabet
//3. The resultant_word is store at each node to make the return easier. 
class Solution {
public:
    class TrieNode{
      public:
        bool word_end=false;
        string res_word;
        TrieNode* children[26] = {NULL};
    };
    /** Initialize your data structure here. */
    TrieNode * root = new TrieNode(); 
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode *curr= root;
        int index;
        for(index=0; index<word.length();index++){
            if(curr->children[word[index]-'a'] ==NULL){
                curr->children[word[index]-'a'] = new TrieNode();
            }
            curr = curr->children[word[index]-'a'];
        }
        curr->word_end = true;
        curr->res_word += word;
    }
    
    string longestWord(vector<string>& words) {
        //edge case
        if(words.size() == 0){
            return "";
        }
        
        //logic
        //construct the trie : O(NxL)
        for(int i=0; i<words.size(); i++){
            insert(words[i]);
        }
        
        //searching for the longest word
        queue<TrieNode *> myqueue;
        myqueue.push(root);
        int size, index;TrieNode* curr;
        while(!myqueue.empty()){
            size = myqueue.size();
            while(size>0){
                curr = myqueue.front(); 
                myqueue.pop(); 
                size--;
                for(int k=25; k>=0;k--){
                    if(curr->children[k] !=NULL && curr->children[k]->word_end == true)     myqueue.push(curr->children[k]);
                }
            }
        }
       return curr->res_word;
    }
};

File: LongestWordDictionary.cs
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
    
    public class TrieNode{
        public TrieNode[] children;
        public string word;
        
        public TrieNode(){
            children = new TrieNode[26];
        }
    }
    
    public void Insert(string word)
    {
        var curr = root;

        for(int i = 0; i < word.Length; i++)
        {
            char c = word[i];
            if(curr.children[c- 'a'] == null)
                curr.children[c-'a'] = new TrieNode();

            curr = curr.children[c-'a'];
        }
        curr.word = word;
    }    
    
    TrieNode root;

    public string LongestWord(string[] words) {
        if(words == null || words.Length == 0)
            return "";
        
        root = new TrieNode();
        foreach(string word in words)
        {
            Insert(word);
        }
        
        //BFS
        Queue<TrieNode> queue = new  Queue<TrieNode>();
        queue.Enqueue(root);
        
        var curr = new TrieNode();
        while(queue.Count > 0)
        {
            curr = queue.Dequeue();
            for(int i = 25; i >= 0; i--)
            {
                if(curr.children[i] != null && curr.children[i].word != null)
                    queue.Enqueue(curr.children[i]);
            }
        }
        
        if(curr.word == null)
            return "";
        return curr.word;
    }
}